lesson4/XOGame.java

```diff
@@ -0,0 +1,170 @@
+            playerTurn();
+            printMap();
+            if (checkWin(DOT_PLAYER)) {
+                System.out.println("YOU WIN!");
+                break;
+            }
+            if (isMapFull()) {
+                System.out.println("  DRAW");
+                break;
+            }
+
+            aiTurn();
+            printMap();
+            if (checkWin(DOT_AI)) {
+                System.out.println("AI WON ((");
+                break;
+            }
+            if (isMapFull()) {
+                System.out.println("  DRAW");
+                break;
+            }
+        }
+        System.out.println("GAME OVER");
+    }
+
+    public static void prepareMap() {
+        map = new char[SIZE][SIZE];
+        for (int i = 0; i < SIZE; i++) {
+            for (int j = 0; j < SIZE; j++) {
+                map[i][j] = DOT_IS_EMPTY;
+            }
+        }
+    }
+
+    public static void printMap() {
+        System.out.print("  ");
+        for (int i = 1; i <= SIZE; i++) {
+            System.out.print(i + " ");
+        }
+        System.out.println();
+        for (int i = 0; i < SIZE; i++) {
+            System.out.print((i + 1 + " "));
+            for (int j = 0; j < SIZE; j++) {
+                System.out.print(map[j][i] + " ");
+            }
+            System.out.println();
+        }
+        System.out.println();
+    }
+
+    public static void playerTurn() {
+        int x, y;
+        do {
+            System.out.println("Enter coordinates: 'x y'");
+            x = sc.nextInt() - 1; //отнимаем в соответствии с эл-тами массива
+            y = sc.nextInt() - 1;
+        } while (!isCellEmpty(x, y));
+        map[x][y] = DOT_PLAYER;
+    }
+
+    public static boolean isCellEmpty(int x, int y) {
+        if (x < 0 || y < 0 || x >= SIZE || y >= SIZE)
+            return false;
+        if (map[x][y] == DOT_IS_EMPTY) {
+            return true;
+        }
+        return false;
+    }
+
+    public static void aiTurn() { //4. *** Доработать ИИ, чтобы он мог блокировать ходы игрока.
+        for (int i = 0; i < SIZE; i++)
+            for (int j = 0; j < SIZE; j++) {
+                if (isCellEmpty(i, j)) {
+                    map[i][j] = DOT_AI;
+                    if (checkWin(DOT_AI)) return;
+                    map[i][j] = DOT_IS_EMPTY;
+                }
+            }
+        for (int i = 0; i < SIZE; i++)
+            for (int j = 0; j < SIZE; j++) {
+                if (isCellEmpty(i, j)) {
+                    map[i][j] = DOT_PLAYER;
+                    if (checkWin(DOT_PLAYER)) {
+                        map[i][j] = DOT_AI;
+                        return;
+                    }
+                    map[i][j] = DOT_IS_EMPTY;
+                }
+            }
+        int x, y;
+        do {
+            x = random.nextInt(SIZE);
+            y = random.nextInt(SIZE);
+        } while (!isCellEmpty(x, y));
+        map[x][y] = DOT_AI;
+        System.out.printf("Ход ИИ: [%d / %d]\n", x + 1, y + 1);
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.